Source/Parser/Functions/IterableJoiningFunction.cs

```diff
@@ -59,7 +59,10 @@ public override bool Evaluate(InterpreterScope scope, out ExpressionBase result)
                 iteratorScope.AssignVariable(predicateParameter, result);
 
                 if (!predicate.Evaluate(iteratorScope, out result))
+                {
+                    result = new ErrorExpression(string.Format("predicate function did not return any value for the predicate input {0}", input.ToString()));
```

Comment:
`result` is expected to be an `ErrorExpression` at this point, so you don't need to replace it with another error expression - especially if the error came from the actual evaluation logic (like a variable not being defined) as this error would be misleading in that case. Here's a couple unit tests you could use to enforce the desired behavior.
```
    [Test]
    public void TestPredicateWithoutReturn()
    {
        Assert.That(Evaluate("any_of([1], (a) { if (a == 2) return 3 })"),
            Is.EqualTo("Function call did not return a value"));
    }

    [Test]
    public void TestPredicateWithUnknownVariable()
    {
        Assert.That(Evaluate("any_of([1], (a) { return b })"),
            Is.EqualTo("Unknown variable: b"));
    }
```